convert2rhel/unit_tests/pkghandler_test.py

```diff
@@ -1387,6 +1398,79 @@ def test_fix_default_kernel_with_no_incorrect_kernel(self):
         self.assertTrue(len(pkghandler.logging.getLogger.warning_msgs) == 0)
 
 
+@pytest.mark.parametrize(
+    ("version1", "version2", "expected"),
+    (
+        ("123-4.fc35", "123-4.fc35", 0),
+        ("123-5.fc35", "123-4.fc35", 1),
+        ("123-3.fc35", "123-4.fc35", -1),
+        ("4.6~pre16262021g84ef6bd9-3.fc35", "4.6~pre16262021g84ef6bd9-3.fc35", 0),
+        ("2:8.2.3568-1.fc35", "2:8.2.3568-1.fc35", 0),
+    ),
+)
+def test_compare_package_versions(version1, version2, expected):
+    assert pkghandler.compare_package_versions(version1, version2) == expected
+
+
+@pytest.mark.parametrize(
+    ("package_manager_type", "packages"),
+    (
+        (
+            "yum",
+            [
+                "convert2rhel.noarch-0.24-1.20211111151554764702.pr356.28.ge9ed160.el8",
+                "convert2rhel.src-0.24-1.20211111151554764702.pr356.28.ge9ed160.el8",
+            ],
+        ),
+        (
+            "dnf",
+            [
+                "dunst-1.7.1-1.fc35.x86_64",
+                "dunst-1.7.0-1.fc35.x86_64",
+                "java-11-openjdk-headless-1:11.0.13.0.8-2.fc35.x86_64",
+            ],
+        ),
+    ),
+)
+def test_get_total_packages_to_update(package_manager_type, packages, monkeypatch):
+    monkeypatch.setattr(pkgmanager, "TYPE", package_manager_type)
+    monkeypatch.setattr(pkghandler, "_get_packages_to_update_%s" % package_manager_type, value=lambda: packages)
+
+    assert get_total_packages_to_update() == len(packages)
+
+
+@pytest.mark.skipif(pkgmanager.TYPE != "yum", reason="No yum module detected on the system, skipping it.")
+@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
+def test_get_packages_to_update_yum(packages, monkeypatch):
+    from cli import YumBaseCli  # pylint: disable=E0401
+
+    class PkgUpdates:
+        def __init__(self, packages):
+            self.updates = packages
+
+    pkg_lists_mock = mock.Mock(return_value=PkgUpdates(packages))
+
+    monkeypatch.setattr(YumBaseCli, "returnPkgLists", value=pkg_lists_mock)
+
+    assert _get_packages_to_update_yum() == packages
+
+
+@pytest.mark.skipif(pkgmanager.TYPE != "dnf", reason="No dnf module detected on the system, skipping it.")
+@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
+def test_get_packages_to_update_yum(packages, monkeypatch):
+    from dnf import Base  # pylint: disable=E0401
+
+    dummy_mock = mock.Mock()
+
+    monkeypatch.setattr(Base, "read_all_repos", value=dummy_mock)
+    monkeypatch.setattr(Base, "fill_sack", value=dummy_mock)
+    monkeypatch.setattr(Base, "upgrade_all", value=dummy_mock)
+    monkeypatch.setattr(Base, "resolve", value=dummy_mock)
+    monkeypatch.setattr(Base, "transaction", value=packages)
+
+    assert _get_packages_to_update_dnf() == packages
```

Comment:
With this `pytest.mark.skipif` our coverage decreases a little because there will not be both of the package managers at the same time on a system. 

I know there is a way to combine multiple coverage runs to make it better, but don't know how much effort will be to adapt our CI to do that.

Reply:
One thing that could be possible is splitting the `test_get_total_packages_to_update` test and reapplying these specific monkeypatchs to be compatible with either dnf or yum when running, so it could be something like this: 

* test_get_total_packages_to_update_with_yum
* test_get_total_packages_to_update_with_dnf

Reply:
Decrease in coverage is fine, code coverage is only a metric - not a goal

You should rename the method though, there are two tests methods with the same name 

```suggestion
@pytest.mark.skipif(pkgmanager.TYPE != "yum", reason="No yum module detected on the system, skipping it.")
@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
def test_get_packages_to_update_yum(packages, monkeypatch):
    from cli import YumBaseCli  # pylint: disable=E0401

    class PkgUpdates:
        def __init__(self, packages):
            self.updates = packages

    pkg_lists_mock = mock.Mock(return_value=PkgUpdates(packages))

    monkeypatch.setattr(YumBaseCli, "returnPkgLists", value=pkg_lists_mock)

    assert _get_packages_to_update_yum() == packages


@pytest.mark.skipif(pkgmanager.TYPE != "dnf", reason="No dnf module detected on the system, skipping it.")
@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
def test_get_packages_to_update_dnf(packages, monkeypatch):
    from dnf import Base  # pylint: disable=E0401

    dummy_mock = mock.Mock()

    monkeypatch.setattr(Base, "read_all_repos", value=dummy_mock)
    monkeypatch.setattr(Base, "fill_sack", value=dummy_mock)
    monkeypatch.setattr(Base, "upgrade_all", value=dummy_mock)
    monkeypatch.setattr(Base, "resolve", value=dummy_mock)
    monkeypatch.setattr(Base, "transaction", value=packages)

    assert _get_packages_to_update_dnf() == packages
```

Reply:
Decrease in coverage is fine, code coverage is only a metric - not a goal

You should rename the method though, there are two tests methods with the same name 

```suggestion
@pytest.mark.skipif(pkgmanager.TYPE != "yum", reason="No yum module detected on the system, skipping it.")
@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
def test_get_packages_to_update_yum(packages, monkeypatch):
    from cli import YumBaseCli  # pylint: disable=E0401

    class PkgUpdates:
        def __init__(self, packages):
            self.updates = packages

    pkg_lists_mock = mock.Mock(return_value=PkgUpdates(packages))

    monkeypatch.setattr(YumBaseCli, "returnPkgLists", value=pkg_lists_mock)

    assert _get_packages_to_update_yum() == packages


@pytest.mark.skipif(pkgmanager.TYPE != "dnf", reason="No dnf module detected on the system, skipping it.")
@pytest.mark.parametrize(("packages"), ((["package-1", "package-2", "package-3"],)))
def test_get_packages_to_update_dnf(packages, monkeypatch):
    from dnf import Base  # pylint: disable=E0401

    dummy_mock = mock.Mock()

    monkeypatch.setattr(Base, "read_all_repos", value=dummy_mock)
    monkeypatch.setattr(Base, "fill_sack", value=dummy_mock)
    monkeypatch.setattr(Base, "upgrade_all", value=dummy_mock)
    monkeypatch.setattr(Base, "resolve", value=dummy_mock)
    monkeypatch.setattr(Base, "transaction", value=packages)

    assert _get_packages_to_update_dnf() == packages
```